text/0000-extension-layer.md

```diff
@@ -0,0 +1,118 @@
+# Meta
+[meta]: #meta
+- Name: Add extension layer to exchange data
+- Start Date: 2023-10-09
+- Author(s): [c0d1ngm0nk3y](https://github.com/c0d1ngm0nk3y), [pbusko](https://github.com/pbusko)
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request:
+- CNB Pull Request:
+- CNB Issue:
+- Related: [RFC#0105 Support Dockerfiles](https://github.com/buildpacks/rfcs/blob/main/text/0105-dockerfiles.md)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This RFC introduces support for Extension Layers to allow data transfer between the build environment and the Kaniko execution.
+
+# Motivation
+[motivation]: #motivation
+
+This change allows extensions to possess their own layers to utilize during the generation/extend process. Additionally, it ensures that extension output does not inadvertently interfere with other extension or buildpack layers during the build, and it does not unintentionally become part of the final application image.
+
+This would allow distroless run images to be extended.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This follows up on RFC-0105 and proposes that during the execution of the extension's `./bin/generate`, an extension is allowed to write arbitrary data to the `context` folder within its exclusive layer. This data then becomes accessible during the execution of the `extend` phase via Kaniko build context. The content of these extension-specific layers is ignored at build and launch time, it serves only the extension phase.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+- New root directory `/layers-ext` is introduced which contains extension layers.
```

Comment:
Is there a better name than `/layers-ext`?

Reply:
Something like `/extensions-gen`?

Reply:
Or perhaps `/contexts`? With apologies for my bikeshedding... buildpacks generate "layers" and extensions generate "contexts" has a nice parity.

Reply:
> buildpacks generate "layers" and extensions generate "contexts" has a nice parity.

The most important piece of information produced by extensions are the Dockerfiles, so IMHO the folder should have somewhat more general name, which covers contexts, Dockerfiles and additional toml files with build arguments. However I do not see it as a deal breaker, so I'm perfectly fine with `/contexts`

Reply:
> the folder should have somewhat more general name, which covers contexts, Dockerfiles and additional toml files with build arguments

This is a fair point. And in fact, we already have `CNB_GENERATED_DIR` which defaults to `<layers>/generated`. Perhaps we can simply copy any output `./context` directories to `<layers>/generated/<extension-id>/<build|run>/context`.

Currently, the lifecycle creates a temp directory for extension output. But perhaps it would be more efficient to provide `<layers>/generated/<extension-id>` as `CNB_OUTPUT_DIR`. This would require some small changes to the Buildpack API for extensions.